--- build/src/tor-0.2.5.8-rc/src/common/compat.c.orig	2014-09-25 15:05:39.000000000 -0400
+++ build/src/tor-0.2.5.8-rc/src/common/compat.c	2014-09-25 15:06:50.000000000 -0400
@@ -1979,16 +1979,16 @@
 #ifdef PR_SET_DUMPABLE
   attempted = 1;
   r = prctl(PR_SET_DUMPABLE, 0);
 #endif
 #endif
-#if defined(__APPLE__) && defined(PT_DENY_ATTACH)
-  if (r < 0) {
-    attempted = 1;
-    r = ptrace(PT_DENY_ATTACH, 0, 0, 0);
-  }
-#endif
+//#if defined(__APPLE__) && defined(PT_DENY_ATTACH)
+//  if (r < 0) {
+//    attempted = 1;
+//    r = ptrace(PT_DENY_ATTACH, 0, 0, 0);
+//  }
+//#endif
 
   // XXX: TODO - Mac OS X has dtrace and this may be disabled.
   // XXX: TODO - Windows probably has something similar
   if (r == 0 && attempted) {
     log_debug(LD_CONFIG,"Debugger attachment disabled for "
