--- build/src/tor-0.2.6.2-alpha/src/common/compat.c.orig	2014-12-30 09:06:12.000000000 -0500
+++ build/src/tor-0.2.6.2-alpha/src/common/compat.c	2015-02-09 23:54:12.000000000 -0500
@@ -2173,28 +2173,28 @@
 #endif
 }
 
-#ifndef HAVE__NSGETENVIRON
+//#ifndef HAVE__NSGETENVIRON
 #ifndef HAVE_EXTERN_ENVIRON_DECLARED
 /* Some platforms declare environ under some circumstances, others don't. */
 #ifndef RUNNING_DOXYGEN
 extern char **environ;
 #endif
 #endif
-#endif
+//#endif
 
 /** Return the current environment. This is a portable replacement for
  * 'environ'. */
 char **
 get_environment(void)
 {
-#ifdef HAVE__NSGETENVIRON
+//#ifdef HAVE__NSGETENVIRON
   /* This is for compatibility between OSX versions.  Otherwise (for example)
    * when we do a mostly-static build on OSX 10.7, the resulting binary won't
    * work on OSX 10.6. */
-  return *_NSGetEnviron();
-#else
+//  return *_NSGetEnviron();
+//#else
   return environ;
-#endif
+//#endif
 }
 
 /** Get name of current host and write it to <b>name</b> array, whose
