--- build/src/tor-0.2.3.19-rc/src/common/compat.c.orig	2012-07-20 12:11:13.000000000 -0700
+++ build/src/tor-0.2.3.19-rc/src/common/compat.c	2012-07-20 12:12:07.000000000 -0700
@@ -1797,32 +1797,32 @@
   }
   return absfname;
 #endif
 }
 
-#ifndef HAVE__NSGETENVIRON
+//#ifndef HAVE__NSGETENVIRON
 #ifndef HAVE_EXTERN_ENVIRON_DECLARED
 /* Some platforms declare environ under some circumstances, others don't. */
 #ifndef RUNNING_DOXYGEN
 extern char **environ;
 #endif
 #endif
-#endif
+//#endif
 
 /** Return the current environment. This is a portable replacement for
  * 'environ'. */
 char **
 get_environment(void)
 {
-#ifdef HAVE__NSGETENVIRON
+//#ifdef HAVE__NSGETENVIRON
   /* This is for compatibility between OSX versions.  Otherwise (for example)
    * when we do a mostly-static build on OSX 10.7, the resulting binary won't
    * work on OSX 10.6. */
-  return *_NSGetEnviron();
-#else
+//  return *_NSGetEnviron();
+//#else
   return environ;
-#endif
+//#endif
 }
 
 /** Set *addr to the IP address (in dotted-quad notation) stored in c.
  * Return 1 on success, 0 if c is badly formatted.  (Like inet_aton(c,addr),
  * but works on Windows and Solaris.)
